File: backend/app/schemas/customer_experience.py
"""Customer experience API schemas."""

from __future__ import annotations

from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class ServiceCategoryRow(BaseModel):
    id: UUID
    slug: str
    name: str
    description: str | None
    icon: str | None
    sort_order: int
    is_active: bool


class FacilityTagRow(BaseModel):
    id: UUID
    slug: str
    name: str
    sort_order: int
    is_active: bool


class ServiceCatalogItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    category: str
    unit: str
    price_inr: Decimal
    description: str | None = None
    estimated_duration_minutes: int | None = None
    express_available: bool = False
    pickup_available: bool = True
    delivery_available: bool = True
    catalog_status: str = "active"
    view_count: int = 0
    order_count: int = 0
    sort_order: int = 0
    is_active: bool = True


class ServiceCatalogCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    category: str = Field(min_length=1, max_length=80)
    unit: str = Field(default="piece", max_length=40)
    price_inr: Decimal = Field(gt=Decimal("0"))
    description: str | None = None
    estimated_duration_minutes: int | None = Field(default=None, ge=5, le=10080)
    express_available: bool = False
    pickup_available: bool = True
    delivery_available: bool = True
    catalog_status: str = Field(default="active", pattern="^(active|paused|draft)$")
    sort_order: int = Field(default=0, ge=0)


class ServiceCatalogUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    category: str | None = Field(default=None, min_length=1, max_length=80)
    unit: str | None = Field(default=None, max_length=40)
    price_inr: Decimal | None = Field(default=None, gt=Decimal("0"))
    description: str | None = None
    estimated_duration_minutes: int | None = Field(default=None, ge=5, le=10080)
    express_available: bool | None = None
    pickup_available: bool | None = None
    delivery_available: bool | None = None
    catalog_status: str | None = Field(default=None, pattern="^(active|paused|draft)$")
    sort_order: int | None = Field(default=None, ge=0)
    is_active: bool | None = None


class ContactInfoResponse(BaseModel):
    offline_booking_mode: bool = False
    can_contact: bool
    requires_login: bool
    show_call: bool
    show_whatsapp: bool
    show_callback: bool
    phone: str | None = None
    whatsapp_number: str | None = None
    whatsapp_url: str | None = None
    address_line: str | None = None
    city: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    full_address: str | None = None
    map_url: str | None = None
    working_hours: dict[str, str] | None = None


class TrackEngagementRequest(BaseModel):
    event_type: str
    service_id: UUID | None = None
    source: str | None = Field(default=None, max_length=40)


class CallbackRequestCreate(BaseModel):
    phone: str = Field(min_length=8, max_length=20)
    preferred_time: str | None = Field(default=None, max_length=120)


class QuestionCreate(BaseModel):
    question: str = Field(min_length=5, max_length=1000)


class QuestionAnswerRequest(BaseModel):
    answer: str = Field(min_length=1, max_length=2000)


class QuestionRow(BaseModel):
    id: UUID
    laundry_id: UUID
    customer_id: UUID
    question: str
    answer: str | None
    status: str
    answered_at: str | None
    created_at: str


class EngagementAnalytics(BaseModel):
    store_views: int
    service_views: int
    calls_generated: int
    whatsapp_clicks: int
    questions_asked: int
    callback_requests: int
    conversion_rate_pct: float


class AdminEngagementOverview(BaseModel):
    store_views: int
    service_views: int
    calls_generated: int
    whatsapp_clicks: int
    questions_asked: int
    callback_requests: int
    pending_questions: int
    pending_storefronts: int

File: frontend/services/customer-experience.ts
import { api, type ApiEnvelope } from '@/lib/api';

export interface ServiceCatalogItem {
  id: string;
  name: string;
  category: string;
  unit: string;
  price_inr: string;
  description?: string | null;
  estimated_duration_minutes?: number | null;
  express_available?: boolean;
  pickup_available?: boolean;
  delivery_available?: boolean;
  catalog_status?: string;
  view_count?: number;
  order_count?: number;
  sort_order?: number;
  is_active: boolean;
}

export interface ServiceCategory {
  id: string;
  slug: string;
  name: string;
  description: string | null;
  icon: string | null;
  sort_order: number;
  is_active: boolean;
}

export interface ContactInfo {
  offline_booking_mode?: boolean;
  contact_available?: boolean;
  can_contact: boolean;
  requires_login: boolean;
  show_call: boolean;
  show_whatsapp: boolean;
  show_callback: boolean;
  phone: string | null;
  whatsapp_number: string | null;
  whatsapp_url: string | null;
  address_line?: string | null;
  city?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  full_address?: string | null;
  map_url?: string | null;
  working_hours?: Record<string, string> | null;
}

export interface CustomerQuestion {
  id: string;
  laundry_id: string;
  customer_id: string;
  question: string;
  answer: string | null;
  status: string;
  answered_at: string | null;
  created_at: string;
}

export interface EngagementAnalytics {
  store_views: number;
  service_views: number;
  calls_generated: number;
  whatsapp_clicks: number;
  questions_asked: number;
  callback_requests: number;
  conversion_rate_pct: number;
}

export type ServiceSort = 'popular' | 'price_asc' | 'price_desc';

export async function listServiceCategories(): Promise<ServiceCategory[]> {
  const { data } = await api.get<ApiEnvelope<ServiceCategory[]>>('/service-categories');
  return data.data;
}

export async function browseServiceCatalog(
  laundryId: string,
  params?: { q?: string; category?: string; express_only?: boolean; sort?: ServiceSort },
): Promise<ServiceCatalogItem[]> {
  const { data } = await api.get<ApiEnvelope<ServiceCatalogItem[]>>(
    `/laundries/${laundryId}/services/catalog`,
    { params },
  );
  return data.data;
}

export async function getServiceDetail(laundryId: string, serviceId: string): Promise<ServiceCatalogItem> {
  const { data } = await api.get<ApiEnvelope<ServiceCatalogItem>>(
    `/laundries/${laundryId}/services/${serviceId}`,
  );
  return data.data;
}

export async function trackStoreView(laundryId: string): Promise<void> {
  await api.post(`/laundries/${laundryId}/engagement/store-view`);
}

export async function getContactInfo(laundryId: string): Promise<ContactInfo> {
  const { data } = await api.get<ApiEnvelope<ContactInfo>>(`/laundries/${laundryId}/contact`);
  return data.data;
}

export async function trackContactEvent(
  laundryId: string,
  body: { event_type: string; service_id?: string; source?: string },
): Promise<ContactInfo> {
  const { data } = await api.post<ApiEnvelope<ContactInfo>>(`/laundries/${laundryId}/contact/track`, body);
  return data.data;
}

export async function requestCallback(
  laundryId: string,
  body: { phone: string; preferred_time?: string },
): Promise<{ id: string; status: string }> {
  const { data } = await api.post<ApiEnvelope<{ id: string; status: string }>>(
    `/laundries/${laundryId}/callback`,
    body,
  );
  return data.data;
}

export async function listPublicQuestions(laundryId: string): Promise<CustomerQuestion[]> {
  const { data } = await api.get<ApiEnvelope<CustomerQuestion[]>>(`/laundries/${laundryId}/questions`);
  return data.data;
}

export async function askQuestion(laundryId: string, question: string): Promise<CustomerQuestion> {
  const { data } = await api.post<ApiEnvelope<CustomerQuestion>>(`/laundries/${laundryId}/questions`, {
    question,
  });
  return data.data;
}

export async function getPartnerEngagementAnalytics(days = 30): Promise<EngagementAnalytics> {
  const { data } = await api.get<ApiEnvelope<EngagementAnalytics>>('/partner/engagement-analytics', {
    params: { days },
  });
  return data.data;
}

export async function listPartnerQuestions(): Promise<CustomerQuestion[]> {
  const { data } = await api.get<ApiEnvelope<CustomerQuestion[]>>('/partner/questions');
  return data.data;
}

export async function answerPartnerQuestion(questionId: string, answer: string): Promise<CustomerQuestion> {
  const { data } = await api.post<ApiEnvelope<CustomerQuestion>>(`/partner/questions/${questionId}/answer`, {
    answer,
  });
  return data.data;
}
